File: 494. Target Sum/Java/Solution.java
public class Solution {
    public int findTargetSumWays(int[] nums, int S) {
        int sumNums = 0;
        for (int num : nums) {
            sumNums += num;
        }
        if (S > sumNums || S < -sumNums) {
            return 0;
        }
        int[] map = new int[2 * sumNums + 1];
        map[sumNums] = 1;
        for (int num : nums) {
            int[] newMap = new int[2 * sumNums + 1];
            for (int i = 0; i < 2 * sumNums + 1; i++) {
                if (map[i] != 0) {
                    newMap[i + num] += map[i];
                    newMap[i - num] += map[i];
                }
            }
            map = newMap;
        }
        return map[sumNums + S];
    }
}

File: 494. Target Sum/Python/Solution.py
class Solution(object):
    def findTargetSumWays(self, nums, S):
        """
        :type nums: List[int]
        :type S: int
        :rtype: int
        """
        if S > 1000:
            return 0
        memory = {}
        memory[nums[0]] = 1
        memory[-nums[0]] = 1 if nums[0] != 0 else 2
        for num in nums[1:]:
            new_memory = {}
            for key, value in memory.items():
                if key + num not in new_memory:
                    new_memory[key + num] = 0
                new_memory[key + num] += value
                if key - num not in new_memory:
                    new_memory[key - num] = 0
                new_memory[key - num] += value
            memory = new_memory
        if S not in memory:
            return 0
        return memory[S]
